### PR TITLE
Refactor Tokenizer -> BaseTokenizer

### PR DESCRIPTION
**Pull request by H-Huang**
This causes breaking changes and users will need to redownload the tokenizer files (`python scripts/download_tokenizer.py ...`)

- Remove `tiktoken` dependency, remove `tiktoken.py`
- Refactor the `Tokenizer` base class
- Update config files to point to directory instead of `tokenizer.model`
- Raise exception if using `tokenizer.model` for tokenizer_path

### CODE REVIEW

File: .ci/docker/requirements.txt [hunk 1]
```diff
@@ -2,7 +2,6 @@ torchdata >= 0.8.0
 datasets >= 3.6.0
 tomli >= 1.1.0 ; python_version < "3.11"
 tensorboard
-tiktoken
```
Comment (tianyu-l) [inline .ci/docker/requirements.txt:5]: please update https://github.com/pytorch/torchtitan/blob/main/pyproject.toml as well

File: .ci/docker/requirements-dev.txt [hunk 2]
```diff
@@ -3,3 +3,4 @@ pytest==7.3.2
 pytest-cov
 pre-commit
 tomli-w >= 1.1.0
+transformers
```
Comment (tianyu-l) [inline .ci/docker/requirements-dev.txt:6]: is this for running the unit tests?

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -14,7 +14,7 @@ We actively welcome your pull requests.
 2. If you've added code that should be tested, add tests.
 3. If you've changed APIs, update the documentation.
 4. Ensure the test suite passes.
-5. Make sure your code lints (`pre-commit run --all-files`).
+5. Make sure your code lints (`pre-commit run --from-ref origin/main --to-ref HEAD`).
```
Comment (tianyu-l) [inline CONTRIBUTING.md]: IIUC this is restricting the linting to be changes between current main and the latest commit. Can I ask why?

File: tests/assets/tokenizer/tokenizer.json
Comment (tianyu-l) [inline tests/assets/tokenizer/tokenizer.json:1]: what's the source of the files under `tests/assets/tokenizer`?
asking because not sure about legal side of things

File: torchtitan/components/tokenizer.py [hunk 4]
```diff
@@ -87,87 +88,79 @@ def _load_tokenizer_from_path(self, tokenizer_path: str) -> HfTokenizer:
...
-
+        # Strategy 1: Load from tokenizer.json (preferred for modern tokenizers)
+        if os.path.exists(tokenizer_json_path):
+            print("Loading tokenizer from tokenizer.json")
+            return Tokenizer.from_file(tokenizer_json_path)
+        # Strategy 2: Load from vocab files (with or without merges.txt)
+        elif os.path.exists(vocab_json_path) or os.path.exists(vocab_txt_path):
+            # Load vocabulary
+            if os.path.exists(vocab_json_path):
+                print("Loading vocabulary from vocab.json")
```
Comment (tianyu-l) [inline torchtitan/components/tokenizer.py]: let's use `logger.info` instead of `print`

File: torchtitan/components/tokenizer.py [hunk 5]
```diff
@@ -87,87 +88,79 @@ def _load_tokenizer_from_path(self, tokenizer_path: str) -> HfTokenizer:
...
-                available_files = [
-                    f
-                    for f in os.listdir(tokenizer_path)
-                    if os.path.isfile(os.path.join(tokenizer_path, f))
-                ]
-                raise FileNotFoundError(
-                    f"No supported tokenizer files found in '{tokenizer_path}'. "
-                    f"Available files: {available_files}. "
-                    "Looking for: tokenizer.json, tokenizer.model, vocab.txt+merges.txt, or vocab.json+merges.txt"
+                print("Loading vocabulary from vocab.txt")
```
Comment (tianyu-l) [inline torchtitan/components/tokenizer.py]: same

File: torchtitan/components/tokenizer.py [hunk 6]
```diff
@@ -87,87 +88,79 @@ def _load_tokenizer_from_path(self, tokenizer_path: str) -> HfTokenizer:
...
+                with open(vocab_txt_path, "r") as f:
+                    for i, line in enumerate(f):
+                        token = line.strip()
+                        if token:
+                            vocab[token] = i
+                vocab_source = "vocab.txt"
+
+            # Strategy 2a: Use BPE if merges.txt exists
+            if os.path.exists(merges_txt_path):
+                print(f"Loading BPE tokenizer from {vocab_source} + merges.txt")
```
Comment (tianyu-l) [inline torchtitan/components/tokenizer.py]: same

File: torchtitan/components/tokenizer.py [hunk 7]
```diff
@@ -87,87 +88,79 @@ def _load_tokenizer_from_path(self, tokenizer_path: str) -> HfTokenizer:
...
 
-        except Exception as e:
-            if isinstance(e, FileNotFoundError):
-                raise e
-            raise Exception(
-                f"Failed to load tokenizer from '{tokenizer_path}': {e}"
-            ) from e
+            # Strategy 2b: Use WordLevel if no merges.txt
+            else:
+                print(f"Loading WordLevel tokenizer from {vocab_source}")
```
Comment (tianyu-l) [inline torchtitan/components/tokenizer.py]: last one

File: torchtitan/experiments/llama4/train_configs/llama4_17bx128e.toml [hunk 8]
```diff
@@ -17,7 +17,7 @@ save_tb_folder = "tb"
 [model]
 name = "llama4"
 flavor = "17bx128e"
-tokenizer_path = "./assets/tokenizer/tokenizer.model"
+tokenizer_path = "./assets/tokenizer/Llama-3.1-8B"
```
Comment (tianyu-l) [inline torchtitan/experiments/llama4/train_configs/llama4_17bx128e.toml]: llama4 tokenizer is different from llama3
See https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/llama4/README.md?plain=1#L14
Please also update the `README.md`

File: torchtitan/experiments/llama4/train_configs/llama4_17bx16e.toml [hunk 9]
```diff
@@ -17,7 +17,7 @@ save_tb_folder = "tb"
 [model]
 name = "llama4"
 flavor = "17bx16e"
-tokenizer_path = "./assets/tokenizer/tokenizer.model"
+tokenizer_path = "./assets/tokenizer/Llama-3.1-8B"
```
Comment (tianyu-l) [inline torchtitan/experiments/llama4/train_configs/llama4_17bx16e.toml]: same

File: torchtitan/experiments/multimodal/tokenizer/tiktoken.py
Comment (tianyu-l) [inline torchtitan/experiments/multimodal/tokenizer/tiktoken.py:1]: We probably should remove / refactor this file too.
But it's out of the scope of this PR.

Let's add a TODO.

File: torchtitan/models/llama3/train_configs/llama3_405b.toml [hunk 10]
```diff
@@ -18,7 +18,7 @@ save_tb_folder = "tb"
 [model]
 name = "llama3"
 flavor = "405B"
-tokenizer_path = "./assets/tokenizer/original/tokenizer.model"
+tokenizer_path = "./assets/tokenizer/meta-llama/Llama-3.1-8B"
```
Comment (tianyu-l) [inline torchtitan/models/llama3/train_configs/llama3_405b.toml]: how come it's different among three llama3 configs

**Review by tianyu-l — COMMENTED** [review]
Looks great in general. Left some comments.

File: .ci/docker/requirements-dev.txt (hunk 2 above)
Comment (H-Huang) [reply .ci/docker/requirements-dev.txt:6]: Yes, it's only used in unit tests

File: CONTRIBUTING.md (hunk 3 above)
Comment (H-Huang) [reply CONTRIBUTING.md]: I needed to update this because when I was running `pre-commit run --all-files`, it includes files not in my pending changes so I was getting a lot of errors for files under `torchtitan/experiments/kernels/`. I updated this it to be `pre-commit run --files $(git diff --name-only HEAD~1)` which is more general

File: tests/assets/tokenizer/tokenizer.json
Comment (H-Huang) [reply tests/assets/tokenizer/tokenizer.json:1]: The old `test_tiktoken.model` had a truncated vocabluary of 2000 based on `tokenizer.model`. Likewise I chatGPTed a python script that takes [tokenizer.json](https://huggingface.co/meta-llama/Llama-3.1-8B/blob/main/tokenizer.json) and reduces it to 2000 tokens.

So this isn't pulled from anywhere, but it is derived from llama-3.1-8B like before, which i assume is okay

File: torchtitan/models/llama3/train_configs/llama3_405b.toml (hunk 10 above)
Comment (H-Huang) [reply torchtitan/models/llama3/train_configs/llama3_405b.toml]: oops, good catch, this is an accident

File: torchtitan/experiments/llama4/train_configs/llama4_17bx128e.toml (hunk 8 above)
Comment (H-Huang) [reply torchtitan/experiments/llama4/train_configs/llama4_17bx128e.toml]: didn't realize, updated it, thanks!

File: tests/assets/tokenizer/tokenizer.json
Comment (tianyu-l) [reply tests/assets/tokenizer/tokenizer.json:1]: Previously the `tokenizer.model` is a fake one taken from some other repo.
What matters is not vocab size, but legal issue -- we can't use llama tokenizer directly due to legal reasons. If your new tokenizer is capable of dealing with `.model` tokenizer, could we keep the old one?

File: CONTRIBUTING.md (hunk 3 above)
Comment (tianyu-l) [reply CONTRIBUTING.md]: > I was getting a lot of errors for files under torchtitan/experiments/kernels/

That sounds very bad. I'm not sure how it escaped the check in https://github.com/pytorch/torchtitan/blob/main/.github/workflows/lint.yaml#L37

I don't think we should change official `CONTRIBUTING.md` to accommodate bad behavior. We should either refactor the code to make sure lint passes in all conditions, or just delete those code. We can do this in separate PRs.

Comment (H-Huang) [reply CONTRIBUTING.md]: > That sounds very bad. I'm not sure how it escaped the check in https://github.com/pytorch/torchtitan/blob/main/.github/workflows/lint.yaml#L37

This only lints the changed files, but yeah not sure how it was checked-in in the first place. Will do that in follow up, seems doable to update

File: tests/assets/tokenizer/tokenizer.json
Comment (wwwjn) [reply tests/assets/tokenizer/tokenizer.json:1]: So we want to keep `test_tiktoken.model` because of CI tests?  I remember CI do has access to network. Can we let integration test to download the tokenizer from Llama repo, and then run the integration test, does this sounds all right?

Comment (H-Huang) [reply tests/assets/tokenizer/tokenizer.json:1]: This PR adds `tests/assets/tokenizer/tokenizer.json` which is functionally the same as `test_tiktoken.model` and I updated the `.toml` and tests to reference this instead

**Review by tianyu-l — APPROVED** [review]
LGTM, let's fix the linting thing in followup PRs.

Also please make sure CI passes / understand the error is not related.

**Comment by H-Huang** [comment]
After https://github.com/pytorch/torchtitan/pull/1343, these changes are causing this test:

```torchrun --nproc_per_node=4 --rdzv_backend c10d --rdzv_endpoint=localhost:0 --local-ranks-filter 0,1,2,3 --role rank --tee 3 -m torchtitan.train --job.config_file ./torchtitan/models/llama3/train_configs/debug_model.toml --model.name llama3_simple_fsdp --training.compile --job.dump_folder artifacts-to-be-uploaded/2d --parallelism.tensor_parallel_degree 2```

to fail. The hypothesis is that embedding layer doesn't have padding and uneven sharding might result in this bug

**Comment by H-Huang** [comment]
Update after changing the test vocab_size from 2002 -> 2000, now the test above is passing

File: torchtitan/models/llama3/train_configs/llama3_405b.toml [hunk 11]
```diff
@@ -18,7 +18,7 @@ save_tb_folder = "tb"
 [model]
 name = "llama3"
 flavor = "405B"
-tokenizer_path = "./assets/tokenizer/original/tokenizer.model"
+tokenizer_path = "./assets/tokenizer/Llama-3.1-8B"
```
Comment (mori360) [inline torchtitan/models/llama3/train_configs/llama3_405b.toml:21]: Path is inconsistent with the download path here:
https://github.com/pytorch/torchtitan/blob/681df506a7639a1ece04960b81d2df11e135b919/README.md?plain=1#L107

Comment (H-Huang) [reply torchtitan/models/llama3/train_configs/llama3_405b.toml:21]: My bad, surprisingly both of these will work:
`python scripts/download_tokenizer.py --repo_id meta-llama/Llama-3.1-8B`
`python scripts/download_tokenizer.py --repo_id meta-llama/Meta-Llama-3.1-8B`

but the README and toml are using different folder names, I will update the README to just specify `meta-llama/Llama-3.1-8B`, please do 
`python scripts/download_tokenizer.py --repo_id meta-llama/Llama-3.1-8B`

Comment (H-Huang) [reply torchtitan/models/llama3/train_configs/llama3_405b.toml:21]: Fix: https://github.com/pytorch/torchtitan/pull/1381
